hazelcast/src/test/java/com/hazelcast/map/MapPartitionIteratorWithQueryTest.java

```diff
@@ -0,0 +1,240 @@
+        final String next = iterator.next();
+        assertEquals(value, next);
+    }
+
+    @Test
+    public void test_Next_Returns_Value_On_NonEmptyPartition_and_HasNext_Returns_False_when_Item_Consumed() throws Exception {
+        String key = generateKeyForPartition(instance, 1);
+        String value = randomString();
+        proxy.put(key, value);
+
+        final Iterator<String> iterator = proxy.iterator(10, 1,
+                new GetValueProjection<String>(), TruePredicate.<String, String>truePredicate());
+        final String next = iterator.next();
+        assertEquals(value, next);
+        boolean hasNext = iterator.hasNext();
+        assertFalse(hasNext);
+    }
+
+    @Test
+    public void test_HasNext_Returns_True_On_NonEmptyPartition() throws Exception {
+        final String key = generateKeyForPartition(instance, 1);
+        final String value = randomString();
+        proxy.put(key, value);
+
+        final Iterator<String> iterator = proxy.iterator(10, 1,
+                new TestProjection(), TruePredicate.<String, String>truePredicate());
+        assertTrue(iterator.hasNext());
+    }
+
+
+    @Test
+    public void test_with_projection_and_true_predicate() throws Exception {
+        for (int i = 0; i < 100; i++) {
+            String key = generateKeyForPartition(instance, 1);
+            proxy.put(key, randomString());
+        }
+        final Iterator<String> iterator = proxy.iterator(10, 1,
+                new TestProjection(), TruePredicate.<String, String>truePredicate());
+
+        final ArrayList<String> projected = collectAll(iterator);
+
+        final Collection<String> actualValues = proxy.values();
+        assertEquals(actualValues.size(), projected.size());
+
+        for (String value : actualValues) {
+            assertTrue(projected.contains("dummy" + value));
+        }
+    }
+
+    @Test
+    public void test_with_projection_and_predicate() throws Exception {
+        final MapProxyImpl<String, Integer> intMap =
+                (MapProxyImpl<String, Integer>) instance.<String, Integer>getMap(randomMapName());
+
+        for (int i = 0; i < 100; i++) {
+            String key = generateKeyForPartition(instance, 1);
+            intMap.put(key, i);
+        }
+
+        final Iterator<Map.Entry<String, Integer>> iterator = intMap.iterator(10, 1,
+                new IdentityProjection<Map.Entry<String, Integer>>(),
+                new EvenPredicate());
+
+        final ArrayList<Map.Entry<String, Integer>> projected = collectAll(iterator);
+        for (Map.Entry<String, Integer> i : projected) {
+            assertTrue(i.getValue() % 2 == 0);
+        }
+
+
+        final Collection<Integer> actualValues = intMap.values();
+        assertEquals(actualValues.size() / 2, projected.size());
+
+
+        for (Entry<String, Integer> e : intMap.entrySet()) {
+            if (e.getValue() % 2 == 0) {
+                assertTrue(projected.contains(e));
+            }
+        }
+    }
```

Comment:
Nitpicking: in such multiline tests it's quite hard to distinguish, where's the division between preparing input data, running the business logic, and asserting.